p4-16/psa/examples/psa-example-parser-error-handling.p4

```diff
@@ -220,6 +220,11 @@ control packet_path_to_bits(out bit<3> packet_path_bits,
     }
     table packet_path_convert {
         key = {
+            // Note that P4Runtime v1.0 does not support P4 programs
+            // with table search key fields that have any type except
+            // bit<W>.  See psa-example-parser-error-handling2.p4 for
```

Comment:
To be accurate, we also support serializable enums and custom types where the underlying type is `bit<W>` even though the P4Runtime spec just says (in the known-limitations section):

> FieldMatch, action Param, and controller packet metadata fields only support unsigned bitstrings, i.e. values of type `bit<W>` (not the more general P4Data).

Reply:
Understood.  Open to a PR on the P4Runtime spec that adds those other supported cases?  I know it doesn't help support _this_ example program, but may as well list everything that we expect to work in the P4Runtime spec, if it is more than what the text implies now.

Reply:
Updated the comment to more accurately reflect P4Runtime v1.0 limitations.

Reply:
Yes a PR to update the P4Runtime spec would be great

Reply:
> To be accurate, we also support serializable enums and custom types where the underlying type is `bit<W>` even though the P4Runtime spec just says (in the known-limitations section):
> 
> > FieldMatch, action Param, and controller packet metadata fields only support unsigned bitstrings, i.e. values of type `bit<W>` (not the more general P4Data).

The support for SerEnum and `type` to typeSpecConverter.cpp is being added via the PSA 1.1 PR: https://github.com/p4lang/p4c/pull/1736

Reply:
Having heard "P4Runtime 1.0 already supports `serEnum` and custom types.", I'd like to investigate `error`.  How do we use `error` as a table key?

P4-16 `error` type does not have serialization defined yet.  A table key requires a key size in `bit<W>` format.  The data plane reads this key for its use.  Therefore, one possible format of the key is a serializable error enum.  For brevity, let's call it `serErrorEnum` similar to use of `serEnum` internally in p4c code.  `serErrorEnum` is a new type for p4c.  An example is provided below.

```
error bit<16> ParserErrors {
    ParseError = 0x0, 
    PacketTooShort = 0x1
}
```
p4c folks can comment on proposed p4c changes and P4Runtime folks can comment on p4runtime implications.

Reply:
Where have you heard this? "P4Runtime 1.0 already supports serEnum and custom types."

There is some truth to it, but maybe not in the way you are thinking.

P4Runtime 1.0 supports serEnum pretty much everywhere, including a table search key field, value_set key field, action parameter, controller header metadata field.  It does this by encoding it in P4Runtime messages the same way it encodes `bit<W>` values -- as binary numbers.

P4Runtime 1.0 also supports arbitrary P4 types, but _only in specific use cases_.  For example, it supports arbitrary P4 typed values as digest message fields, or as register array elements.  P4Runtime 1.0 _does not support_ arbitrary P4 types as any of the above 4 things (table search key fields, etc.).  That is the reason there is a specific note in Section 20 of the spec that says so.

The method for serializing `enum` types that are not serializable enums (but only for the cases I mentioned above where they are supported, like digest messages and register array elements) is to encode them in P4Runtime messages as strings, which are the names of the `enum` members from the P4 source code.  Even though they are likely encoded as small integers in the data plane, they are _not_ encoded as integers in P4Runtime messages.  It is the responsibility of the P4runtime server in such cases to translate between the target-specific small integer values in the data plane, and the strings used in P4Runtime messages.  That translation can be different from one target to another.  The P4Info file does _not_ contain the name-to-number mapping for `enum` types that are not also serializable enums, because it is target-specific, and the P4Info file is intended to be the same across all targets.

Reply:
Here is a v1model architecture program with a digest, where the digest message contains a field of a non-serializable enum type, and also a field of type error: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/v1model-grab-bag.p4

And here is the text format of the P4Info file created via the following command, using latest p4c as of 2019-Feb-26:
```
$ p4c-bm2-ss v1model-grab-bag.p4 --p4runtime-files v1model-grab-bag.p4.p4.info.txt
```
https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/v1model-grab-bag.p4.p4info.txt

Here is the part of that file defining a struct type `test_digest_t` that has the enum and error type fields: https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/v1model-grab-bag.p4.p4info.txt#L199-L228

Immediately below those lines you see the P4Info messages that define the names of the members of type error, and the names of the members of type `MyPacketTypes`.  There are no integer encodings, because they are target-specific and thus do not belong in the P4Info file, by design.

Reply:
In this Issue's note from Antonin above as follows:
"To be accurate, we also support serializable enums and custom types where the underlying type is bit<W> even though the P4Runtime spec just says (in the known-limitations section):"

Thanks for the clarification about `type` not used as a table key.  However, I am not interested in `type` used as a table key.   My main focus if to use new `serErrorEnum` as table key. 

Regarding translation from data plane integers to string, for sending string between p4runtime client and server, is a topic you explained yesterday - thanks.  If a string is target-specific, it's included in the binary blob. 

> Where have you heard this? "P4Runtime 1.0 already supports serEnum and custom types."
> 
> There is some truth to it, but maybe not in the way you are thinking.
> 
> P4Runtime 1.0 supports serEnum pretty much everywhere, including a table search key field, value_set key field, action parameter, controller header metadata field. It does this by encoding it in P4Runtime messages the same way it encodes `bit<W>` values -- as binary numbers.
> 
> P4Runtime 1.0 also supports arbitrary P4 types, but _only in specific use cases_. For example, it supports arbitrary P4 typed values as digest message fields, or as register array elements. P4Runtime 1.0 _does not support_ arbitrary P4 types as any of the above 4 things (table search key fields, etc.). That is the reason there is a specific note in Section 20 of the spec that says so.
> 
> The method for serializing `enum` types that are not serializable enums (but only for the cases I mentioned above where they are supported, like digest messages and register array elements) is to encode them in P4Runtime messages as strings, which are the names of the `enum` members from the P4 source code. Even though they are likely encoded as small integers in the data plane, they are _not_ encoded as integers in P4Runtime messages. It is the responsibility of the P4runtime server in such cases to translate between the target-specific small integer values in the data plane, and the strings used in P4Runtime messages. That translation can be different from one target to another. The P4Info file does _not_ contain the name-to-number mapping for `enum` types that are not also serializable enums, because it is target-specific, and the P4Info file is intended to be the same across all targets.

Reply:
I have already said several times, I know the p4runtime 1.0 spec specifies using human-readable text for `error` and `enum`.  If a table key uses a new `serErrorEnum`, the key uses integers, but p4c API generation changes the integers to string representation.

Reply:
OK, then I am unfamiliar with what `serErrorEnum` is, probably because of my ignorance of the compiler internals.  If it is some p4c internal representation of a P4_16 value of type `error`, then it cannot be used as a table search key for P4Runtime v1.0.  Full stop.  It cannot.  Don't try.

Extend P4Runtime v1.0 somehow so it can be used as a table search key is possible.  Trying to cram it into P4Runtime v1.0 is a job for Don Quixote.

Reply:
And if you want to include an `enum` or `error` field in a digest message, it already works (or, at least it already successfully generates a correct-looking P4Info file -- other parts of the implementation may be incomplete), as I demonstrated with the example  P4_16 v1model architecture program I linked to in a comment earlier today.